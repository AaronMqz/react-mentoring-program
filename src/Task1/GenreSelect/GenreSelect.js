import React, { useState } from "react";
import { GenreStyled } from "./GenreSelect.styled";

const GenreSelect = ({ genreList, currentSelected, onSelect }) => {
  const { Container, Content, Label } = GenreStyled;
  const [genre, setGenre] = useState(currentSelected); // ** Maybe this is not needed for task 1, just added to highlight the current gender

  return (
    <Container>
      <Content>
        {genreList.map((item, index) => {
          return (
            <Label
              key={index}
              id={item}
              active={genre === item ? true : false}
              onClick={(e) => {
                setGenre(e.target.id); // ** Maybe this is not needed for task 1, just added to highlight the current gender
                onSelect(e.target.id);
              }}
            >
              {item.toUpperCase()}
            </Label>
          );
        })}
      </Content>
    </Container>
  );
};

export default GenreSelect;
