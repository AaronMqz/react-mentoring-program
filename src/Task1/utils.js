export const GenresList = ["ALL", "DOCUMENTARY", "COMDEY", "HORROR", "CRIME"];
