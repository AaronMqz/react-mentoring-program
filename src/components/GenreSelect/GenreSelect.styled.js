import styled from "styled-components";

const Container = styled.div`
  display: flex;
  justify-content: center;
  padding-top: 15px;
  border-bottom: solid 3px #555555;
`;

const Content = styled.div`
  display: flex;
  height: 40px;
  align-items: flex-end;
`;

const Label = styled.label`
  margin-right: 15px;
  bottom: -3px;
  position: relative;
  padding-bottom: 15px;
  margin-bottom: 0;
  cursor: pointer;
  &:hover {
    border-bottom: solid 3px #f65261;
  }
  ${(props) =>
    props.active
      ? `border-bottom: solid 3px #f65261`
      : `border-bottom: solid 3px #555555`}
`;

export const GenreStyled = {
  Label,
  Content,
  Container,
};
