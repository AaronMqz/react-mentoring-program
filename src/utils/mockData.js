export const mockData = {
  image_url: "https://m.media-amazon.com/images/I/71eHZFw+GlL._AC_SY879_.jpg",
  movie_name: "Avengers: Inifinity War",
  release_year: "2018",
  genres: ["Action", "Adventure"],
  duration: "2h 29min",
  description:
    "As the Avengers and their allies have continued to protect the world from threats too large for any one hero to handle, a new danger has emerged from the cosmic shadows: Thanos. A despot of intergalactic infamy, his goal is to collect all six Infinity Stones, artifacts of unimaginable power, and use them to inflict his twisted will on all of reality. Everything the Avengers have fought for has led up to this moment, the fate of Earth and existence has never been more uncertain.",
  rating: "8.4",
};

export const SortByOptions = [
  { id: "release_date", name: "RELEASE DATE" },
  { id: "title", name: "TITLE" },
];
