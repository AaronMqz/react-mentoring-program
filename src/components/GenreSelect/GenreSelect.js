import React, { useState } from "react";
import { GenreStyled } from "./GenreSelect.styled";

const GenreSelect = ({ genreList, currentSelected, onChange }) => {
  const { Container, Content, Label } = GenreStyled;
  const [genre, setGenre] = useState(currentSelected); // ** Maybe this is not needed for task 1, just added to highlight the current gender

  return (
    <Container>
      <Content>
        {genreList.map((item, index) => {
          return (
            <Label
              aria-label="genre-list"
              key={index}
              id={item}
              active={genre === item ? true : false}
              onClick={(e) => {
                setGenre(e.target.id); // ** Maybe this is not needed for task 1, just added to highlight the current gender
                onChange(e.target.id);
              }}
            >
              {item.toUpperCase()}
            </Label>
          );
        })}
      </Content>
    </Container>
  );
};

export default GenreSelect;
