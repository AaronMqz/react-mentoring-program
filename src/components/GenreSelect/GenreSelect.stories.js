import { within, userEvent } from "@storybook/testing-library";
import GenreSelect from "../../components/GenreSelect/GenreSelect";
import { GenresList } from "../../utils";

// More on how to set up stories at: https://storybook.js.org/docs/7.0/react/writing-stories/introduction
export default {
  title: "Example/GenreSelect",
  component: GenreSelect,
  tags: ["autodocs"],
  argTypes: { onChange: { action: "Genre Selecteed" } },
};

// More on writing stories with args: https://storybook.js.org/docs/7.0/react/writing-stories/args
export const Default = {
  args: {
    genreList: GenresList,
    currentSelected: "ALL",
  },
};

export const SelectGenre = {
  args: {
    genreList: GenresList,
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const actionSelect = await canvas.getByText("CRIME");
    await userEvent.click(actionSelect);
  },
};
