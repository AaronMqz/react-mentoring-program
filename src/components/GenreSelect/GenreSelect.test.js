import React from "react";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import GenreSelect from "./GenreSelect";
import { GenresList } from "../../utils";
import renderer from "react-test-renderer";

afterEach(cleanup);

describe("GenreSelect Tests:", () => {
  test("1. That component renders all genres passed in props.", () => {
    render(<GenreSelect genreList={GenresList} />);
    const list = screen.getAllByLabelText("genre-list");
    const genres = list.map((i) => i.textContent);
    expect(genres).toEqual(GenresList);
  });

  test("2. That component highlights a selected genre passed in props.", () => {
    render(<GenreSelect genreList={GenresList} currentSelected={"HORROR"} />);
    const element = screen.getByText("HORROR");
    const styles = getComputedStyle(element);
    const styleActive = "3px solid #f65261"; // GenreSelect.styled.js - line 28
    expect(styles.borderBottom).toBe(styleActive);
  });

  test("3. That after a click event on a genre button component calls 'onChange' callback and passes correct genre in arguments", () => {
    const mockOnClick = jest.fn();

    render(
      <GenreSelect
        genreList={GenresList}
        currentSelected={"HORROR"}
        onChange={(selected) => mockOnClick(selected)}
      />
    );
    const element = screen.getByText("CRIME");
    fireEvent.click(element);

    expect(mockOnClick).toHaveBeenCalledWith("CRIME");
  });

  it("4. Snapshot renders correctly", () => {
    const tree = renderer
      .create(<GenreSelect genreList={GenresList} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
