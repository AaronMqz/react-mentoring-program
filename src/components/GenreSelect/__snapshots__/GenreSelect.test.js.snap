// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreSelect Tests: 4. Snapshot renders correctly 1`] = `
<div
  className="sc-beqWaB bqHPCd"
>
  <div
    className="sc-gueYoa kerPiR"
  >
    <label
      aria-label="genre-list"
      className="sc-dmqHEX hZEnJb"
      id="ALL"
      onClick={[Function]}
    >
      ALL
    </label>
    <label
      aria-label="genre-list"
      className="sc-dmqHEX hZEnJb"
      id="DOCUMENTARY"
      onClick={[Function]}
    >
      DOCUMENTARY
    </label>
    <label
      aria-label="genre-list"
      className="sc-dmqHEX hZEnJb"
      id="COMDEY"
      onClick={[Function]}
    >
      COMDEY
    </label>
    <label
      aria-label="genre-list"
      className="sc-dmqHEX hZEnJb"
      id="HORROR"
      onClick={[Function]}
    >
      HORROR
    </label>
    <label
      aria-label="genre-list"
      className="sc-dmqHEX hZEnJb"
      id="CRIME"
      onClick={[Function]}
    >
      CRIME
    </label>
  </div>
</div>
`;
